[ABI] Distinguish @objc from Swift protocol descriptor references

DougGregor opened the pull request:
Introduce `TargetProtocolDescriptorRef` for the runtime, which provides
a reference to either a Swift or an Objective-C protocol, using a spare bit
to indicate which kind of protocol it is. This is staging for separating
the runtime layout of Swift protocols from @objc protocols.

`TargetProtocolDescriptorRef` shows up in a few places in the runtime
and it's interaction with IRGen:

* `swift_getExistentialTypeMetadata()`, which passes an array of `TargetProtocolDescriptorRef`s, and `TargetExistentialTypeMetadata`, which stores a copy of that array.
* `TargetGenericRequirement`, for conformance constraints in generic requirement lists

DougGregor committed:
[ABI] Introduce TargetProtocolDescriptorRef.

TargetProtocolDescriptorRef provides a reference to either a Swift or
an Objective-C protocol, using a spare bit to indicate which kind of
protocol it is. At present, the protocol descriptor also has this
information (so this is redundant), but that will change shortly.

There are no clients of TargetProtocolDescriptorRef yet.

DougGregor committed:
[ABI] Remove the “final” assertion from the ABI’s TrailingObjects.

In our ABI, we’re using some C++ ABI tricks to simplify the 
implementation of the runtime. One of them, in FullMetadata, requires
that we inherit from classes that we would also like to make use
TrailingObjects. Remove the “final” assertion from TrailingObjects
so we can do so.

We promise to be careful.

DougGregor committed:
[ABI] Rework existential type metadata to use ProtocolDescriptorRef.

Use ProtocolDescriptorRefs within the runtime representation of
existential type metadata (TargetExistentialTypeMetadata) instead of
bare protocol descriptor pointers. Start rolling out the use of
ProtocolDescriptorRef in a few places in the runtime that touch this
code. Note that we’re not yet establishing any strong invariants on
the TargetProtocolDescriptorRef instances.

While here, replace TargetExistentialTypeMetadata’s hand-rolled pointer 
arithmetic with swift::ABI::TrailingObjects and centralize knowledge of
its layout better.

DougGregor committed:
[Runtime/IRGen] Switch swift_getExistentialTypeMetadata() to ProtocolDescriptorRef.

Switch one entry point in the runtime (swift_getExistentialTypeMetadata)
to use ProtocolDescriptorRef rather than a protocol descriptor. Update
IRGen to produce ProtocolDescriptorRef instances for its calls, setting
the discriminator bit appropriately.

Within the runtime, verify that all instances of ProtocolDescriptorRef have
the right layout, i.e., the discriminator bit is set for @objc protocols
but not Swift protocols.

DougGregor commented:
@swift-ci please test

slavapestov commented on include/swift/ABI/Metadata.h:
```diff
@@ -1577,6 +1577,126 @@ using TupleTypeMetadata = TargetTupleTypeMetadata<InProcess>;
+    }
+#endif
+
+    return getProtocolDescriptorUnchecked()->Flags.needsWitnessTable();
```
Now that AnyObject is gone, will this ever be false for Swift protocols? Or just maybe in the future?

DougGregor replied:
I don't think it will ever be false for Swift protocols, and we could probably drop the flag entirely once I've finished with this pull request. I don't think we'd go the `AnyObject` route again, with a magic protocol that has no witness table... we'd likely add a new kind of generic requirement or flag bit instead.

DougGregor replied:
FYI, `needsWitnessTable` is computed based on the dispatch strategy, so it's always `true` for Swift protocols. I've turned this into a `return true;` in a follow-up PR

slavapestov commented on include/swift/ABI/Metadata.h:
```diff
@@ -1771,26 +1921,53 @@ enum class ExistentialTypeRepresentation {
-struct TargetExistentialTypeMetadata : public TargetMetadata<Runtime> {
+struct TargetExistentialTypeMetadata
+  : TargetMetadata<Runtime>,
+    swift::ABI::TrailingObjects<
```
This made things much cleaner, thanks!

DougGregor committed:
[Runtime] TargetProtocolDescriptorRef cleanups.

Eliminate an unnecessary nullptr constructor and make the Metadata
unit test compile when Objective-C interoperability is disabled.

DougGregor commented:
@swift-ci please test Linux

DougGregor commented:
@swift-ci please test Linux

DougGregor committed:
[ABI] Distinguish Swift/ObjC protocols in TargetGenericRequirement.

In a generic requirement, distinguish between Swift and
Objective-C protocols using a spare bit within the relative
(indirectable) reference to the protocol.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test
